### PR TITLE
Add ssl

### PR DESCRIPTION
**Pull request by axos88**
- add new resource for tls verification with pluggable server setup (implemented for nginx for reference)
  - fixed kitchen tests #53
  - minor cleanups
  - add alt_names for the ssl_selfsigned resource

See #54

### CODE REVIEW

**Comment by axos88** [comment]
ubuntu-1604 will fail, due to:

W: GPG error: http://ftp.igh.cnrs.fr/pub/mariadb/repo/10.1/ubuntu xenial InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY F1656F24C74CD1D8
W: The repository 'http://ftp.igh.cnrs.fr/pub/mariadb/repo/10.1/ubuntu xenial InRelease' is not signed.
N: Data from such a repository can't be authenticated and is therefore potentially dangerous to use.
N: See apt-secure(8) manpage for repository creation and user configuration details.

**Comment by axos88** [comment]
@thoutenbos

**Comment by axos88** [comment]
@thoutenbos

File: .kitchen.yml [hunk 1]
```diff
@@ -22,13 +22,17 @@ driver:
   name: vagrant
   customize:
     memory: 1024
+  network:
+    - ["private_network", {ip: "33.33.33.200"}]
```
Comment (thoutenbos) [inline .kitchen.yml]: What's this for ?

**Comment by thoutenbos** [comment]
I'm still missing the updates to the README on how to use the new provider.
Could you add it ?

**Comment by axos88** [comment]
I will add it, sorry I missed that.

File: .kitchen.yml (hunk 1 above)
Comment (axos88) [inline .kitchen.yml]: I used it during development so I could get direct access to the machines, forgot to remove it. It doesn't hurt that it's there, but I can remove it, if you'd prefer.

Comment (thoutenbos) [inline .kitchen.yml]: I would prefer to remove it indeed, as it is public IP space.

**Comment by axos88** [comment]
@thoutenbos

**Comment by thoutenbos** [comment]
Thanks!

**Comment by axos88** [comment]
You're very welcome. Can you make a release with this please?

**Comment by notapatch** [comment]
Hi @axos88 

I was reading about the [two different challenges - http-01 & tls-sn-01][1] and wondering about the differences. From the end user's perspective, is it about the choice of being challenged over port 80 vs 443? That they are otherwise no more or less secure? Or is there a good reason to choose the tls-sn-01 over http-01?

I ask because most people want the goal of HTTPS but don't want to think too much about different options *unless they have to*. I will then update the README accordingly.

[1]: https://tools.ietf.org/html/draft-ietf-acme-acme-03#section-7

**Comment by axos88** [comment]
The HTTP challange requires a file with specific content to be put into the webroot of the specific site. The TLS challange requires a more-or-less randomly-generated site to be configured with a specific self-signed certificate to complete it.

The problem with the HTTP challange arises when there is no webroot to serve static files from. (in case of forward proxying to a java process, or a ruby on rails process server, etc).

In such cases chef would need to know the intricates of whoever is using the site on order to be able to convince it so serve the challange's static file. 

In case of TLS it can be avoided, as interacting with nginx alone is enough, and chef can be agnostic on what processes will actually use the certificate in the end as there is no collision between the virtualserver name and the one required by the challenge.

**Comment by notapatch** [comment]
Thanks for the reply @axos88 

So, I have configured letsencrypt using http-01 (as it happens for a ruby on rails process) and I had to hard-code a root into the well-known location before it would work (annoying as different versions of NGINX has different root paths). The relevant parts of my configuration are as follows:

    server {
      listen 80;
      server_name example.com;
      return 301 https://$host$1$request_uri;
    }

    server {
      listen 443;
      ssl on;
      ssl_certificate /etc/ssl/certs/example.com.crt;
      ssl_certificate_key /etc/ssl/private/example.com.key;
      server_name example.com;
      root /home/deployer/apps/my_application_production/current/public;
      ...
      ....
      location ~ /.well-known {                    <=== letsencrypt specfic configuration
        root   /var/www/nginx-default;         <=== path to root changes between NGINX versions
      }
    }

 If I had used TLS it would create a virtual block on the fly and my NGINX server block configuration would not have to know about letsencrypt?

If I understand correctly, that does sound cleaner.

**Comment by axos88** [comment]
Exactly.

**Pull request merged**